webmin/webmin-lib.pl

```diff
@@ -108,6 +108,39 @@ =head1 webmin-lib.pl
 our $password_change_mod = "passwd";
 our $password_change_path = "/".$password_change_mod."/change_passwd.cgi";
 
+if (!defined($gconfig{'noselfwebminup'})) {
+	&has_repos();
+	}
+
+=head2 has_repos
+
+Checks if package manager repositories are
+available for Webmin and Usermin updates.
+
+=cut
+sub has_repos
+{
+my ($force) = @_;
+my $has_repos = 0;
+if (-d '/etc/apt' || -d '/etc/yum.repos.d') {
+	my $out = &backquote_command(
+		"grep -irE 'webmin\.com|webmin\.dev|virtualmin\.com|virtualmin\.dev' ".
```

Comment:
Instead of shelling out to the grep command, it's more efficient to do the reading and parsing of these files in Perl

Reply:
> Instead of shelling out to the grep command, it's more efficient to do the reading and parsing of these files in Perl

I know but sadly it's way to much code. Though here is the patch https://github.com/webmin/webmin/commit/575c1ca5a3c90049a48aacfe389994dc6305b96a.